src/concepts/02-events-and-handlers.md

```diff
@@ -0,0 +1,204 @@
+
+```python
+@engine.on(CustomEvents.CUSTOM_STARTED)
+def call_on_custom_event(engine):
+     # do something
+
+@engine.on(Events.STARTED)
+def fire_custom_events(engine):
+     engine.fire_event(CustomEvents.CUSTOM_STARTED)
+```
+
+{{<info "Note:">}}
+See the source code of [`create_supervised_tbptt_trainer()`](https://pytorch.org/ignite/contrib/engines.html#ignite.contrib.engines.tbptt.create_supervised_tbptt_trainer) for an example of usage of custom events.
+{{</info>}}
+
+If you want to use filtering with custom events (e.g.
+`CustomEvents.CUSTOM_STARTED(every=5)`), you need to do 3 more things:
+
+1.   `engine.state` should have corresponding attributes for the events, e.g. `engine.state.custom_started`
+2.   You need to pass a dict *event_to_attr* to [`register_events`](https://pytorch.org/ignite/contrib/engines.html#ignite.contrib.engines.tbptt.create_supervised_tbptt_trainer), which maps between events and state attributes, e.g.
+
+```python
+event_to_attr = {
+    CustomEvents.CUSTOM_STARTED: "custom_started",
+    CustomEvents.CUSTOM_COMPLETED: "custom_completed",
+}
+```
+
+3. You should increase the counter for the event whenever you fire the event, e.g. `engine.state.custom_started += 1`
+
+{{<danger "Warning:">}}
+This solution for filtering is a temporary workaround and may change in the future.
+{{</danger>}}
+
+## Handlers
+
+Ignite provides a set of built-in handlers to checkpoint the training pipeline, to save best models, to stop training if no improvement, to use experiment tracking system, etc. They can be found in the following two modules:
+
+-   [`ignite.handlers`](https://pytorch.org/ignite/handlers.html)
+-   [`ignite.contrib.handlers`](https://pytorch.org/ignite/contrib/handlers.html)
+
+Some classes can be simply added to [`Engine`](https://pytorch.org/ignite/generated/ignite.engine.engine.Engine.html#ignite.engine.engine.Engine) as a callable function. For example,
+
+```python
+from ignite.handlers import TerminateOnNan
+
+trainer.add_event_handler(Events.ITERATION_COMPLETED, TerminateOnNan())
+```
+
+Others provide an `attach` method to internally add several handlers to [`Engine`](https://pytorch.org/ignite/generated/ignite.engine.engine.Engine.html#ignite.engine.engine.Engine):
+
+```python
+from ignite.contrib.handlers.tensorboard_logger import *
+
+# Create a logger
+tb_logger = TensorboardLogger(log_dir="experiments/tb_logs")
+
+# Attach the logger to the trainer to log model's weights as a histogram after each epoch
+tb_logger.attach(
+    trainer,
+    event_name=Events.EPOCH_COMPLETED,
+    log_handler=WeightsHistHandler(model)
+)
+```
+
+## Timeline and events
+
+Below the events and some typical handlers are displayed on a timeline for a training loop with evaluation after every epoch:
+
+![png](/_images/timeline_and_events.png)
```

Comment:
Let's rework this image, https://github.com/pytorch/ignite/issues/2282
In dark mode it is not readable